### PR TITLE
Numpy Trace

### PR DESCRIPTION
**Pull request by hzfan**
## Description ##
Numpy-compatible version of trace with a new kernel

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] numpy-compatible trace operator on CPU and GPU
- [x] unit tests on CPU

## Comments ##
- not tested on GPU

@haojin2 @reminisce

### CODE REVIEW

File: src/operator/numpy/np_trace_op-inl.h [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+#include "../tensor/broadcast_reduce_op.h"
+
+namespace mxnet {
+namespace op {
+
+struct NumpyTraceParam: public dmlc::Parameter<NumpyTraceParam> {
+    int offset, axis1, axis2;
+    DMLC_DECLARE_PARAMETER(NumpyTraceParam) {
+        DMLC_DECLARE_FIELD(offset)
+            .set_default(0)
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: 2-space indentation is preferred. Please fix all of the 4-space indentations in C++ files.

File: src/operator/numpy/np_trace_op-inl.h [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+      } else if (req == kAddTo) {
+        for (index_t k = 0; k < dlength; ++k) {
+          out[i] += a[j];
+          j += stride;
+        }
+      }
+    }
+  }
+};
+
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: nit: delete this extra blank line.

File: src/operator/numpy/np_trace_op-inl.h [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+template<typename xpu, bool back>
+void NumpyTraceOpProcess(const TBlob& in_data,
+                   const TBlob& out_data,
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: please align this line with the argument in the line above. like:
```c++
void NumpyTraceOpProcess(const TBlob& in_data,
                         const TBlob& out_data,
```

Comment (haojin2) [reply src/operator/numpy/np_trace_op-inl.h]: Same for all arguments below and for all other function signatures.

File: src/operator/numpy/np_trace_op-inl.h [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+                   const std::vector<OpReqType>& req) {
+  using namespace mxnet_op;
+  using namespace mshadow;
+  if (dsize == 0) {
+    if (back) {
+      if (out_data.Size() != 0) {
+        MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+          MXNET_ASSIGN_REQ_SWITCH(req[0], req_type, {
+            if (req_type == kWriteTo)
+              out_data.FlatTo1D<xpu, DType>(s) = 0;
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: please do
```c++
if (...) {
  // your code here
}
```
even for 1-line if-statements.

File: src/operator/numpy/np_trace_op-inl.h [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        });
+      }
+    }
+    return;
+  } else if (ishape.Size() == 0) {
+    if (!back) {
+      MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+        MXNET_ASSIGN_REQ_SWITCH(req[0], req_type, {
+          if (req_type == kWriteTo)
+            out_data.FlatTo1D<xpu, DType>(s) = 0;
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: same here.

File: src/operator/numpy/np_trace_op-inl.h [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+  // number of elements in the offset diagonal
+  // may be negative
+  int dlength;
+  if (param.offset > 0)
+      dlength = std::min(ishape[x1], ishape[x2] - param.offset);
+  else if (param.offset < 0)
+      dlength = std::min(ishape[x1] - (-param.offset), ishape[x2]);
+  else
+      dlength = std::min(ishape[x1], ishape[x2]);
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: same here:
```c++
if () {
} else if () {
} else {
}
```

File: src/operator/numpy/np_trace_op-inl.h (hunk 1 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op-inl.h (hunk 2 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op-inl.h (hunk 3 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op-inl.h (hunk 4 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op-inl.h (hunk 5 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op-inl.h (hunk 6 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op-inl.h [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+};
+
+template<int ndim, int req, bool back>
+struct numpy_trace {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(index_t i, DType* out, const DType* a,
+                                  mshadow::Shape<ndim> oshape,
+                                  mshadow::Shape<ndim> ishape,
+                                  index_t stride, index_t offset, int dlength) {
+  using namespace mxnet_op;
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: indent the whole function body

File: src/operator/numpy/np_trace_op-inl.h [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+  const TBlob& in_data = inputs[0];
+  const TBlob& out_data = outputs[0];
+  const mxnet::TShape& ishape = inputs[0].shape_;
+  const mxnet::TShape& oshape = outputs[0].shape_;
+  const NumpyTraceParam& param = nnvm::get<NumpyTraceParam>(attrs.parsed);
+
+  NumpyTraceOpProcess<xpu, false>(in_data, out_data, ishape, oshape,
+                                  out_data.Size(), param, s, req);
+}
+
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: nit: Remove extra blank line here.

File: src/operator/numpy/np_trace_op-inl.h [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+  const mxnet::TShape& oshape = outputs[0].shape_;
+  const NumpyTraceParam& param = nnvm::get<NumpyTraceParam>(attrs.parsed);
+
+  NumpyTraceOpProcess<xpu, true>(in_data, out_data, oshape, ishape,
+                                 in_data.Size(), param, s, req);
+}
+
+}  // namespace op
+}  // namespace mxnet
+
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op-inl.h]: nit: remove extra blank line here.

File: src/operator/numpy/np_trace_op.cc [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    return std::vector<std::string>{"data"};
+  })
+.set_attr<mxnet::FInferShape>("FInferShape", NumpyTraceOpShape)
+.set_attr<nnvm::FInferType>("FInferType", ElemwiseType<1, 1>)
+.set_attr<FCompute>("FCompute<cpu>", NumpyTraceOpForward<cpu>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseNone{"_backward_np_trace"})
+.add_argument("data", "NDArray-or-Symbol", "Input ndarray")
+.add_arguments(NumpyTraceParam::__FIELDS__());
+
+
```
Comment (haojin2) [inline src/operator/numpy/np_trace_op.cc]: nit: remove extra blank line here.

File: tests/python/unittest/test_numpy_op.py [hunk 11]
```diff
@@ -948,6 +948,85 @@ def hybrid_forward(self, F, x):
...
+@with_seed()
+@npx.use_np_shape
+def test_np_trace():
+  class TestTrace(HybridBlock):
+    def __init__(self, axis1, axis2, offset):
+      super(TestTrace, self).__init__()
+      self._axis1 = axis1
+      self._axis2 = axis2
+      self._offset = offset
+      # necessary initializations
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: Remove this comment.

File: tests/python/unittest/test_numpy_op.py [hunk 12]
```diff
@@ -948,6 +948,85 @@ def hybrid_forward(self, F, x):
...
+
+  shapes = [(3, 3),
+            (3, 4),
+            (0, 0),
+            (3, 3, 3),
+            (0, 0, 0),
+            (2, 2, 4, 3),
+            (2, 2, 4, 3),
+            (2, 0, 3, 0),
+            (2, 0, 2, 3)]
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]:
```python
shapes = [
    # shapes here
]
```

File: tests/python/unittest/test_numpy_op.py [hunk 13]
```diff
@@ -948,6 +948,85 @@ def hybrid_forward(self, F, x):
             assert same(ret_mx.asnumpy(), ret_np)
 
 
+@with_seed()
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: Use 4-space indentations in Python files.

File: src/operator/numpy/np_trace_op-inl.h (hunk 7 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op-inl.h (hunk 8 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op-inl.h (hunk 9 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op-inl.h]: fixed

File: src/operator/numpy/np_trace_op.cc (hunk 10 above)
Comment (hzfan) [reply src/operator/numpy/np_trace_op.cc]: fixed

File: tests/python/unittest/test_numpy_op.py (hunk 11 above)
Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: fixed

File: tests/python/unittest/test_numpy_op.py (hunk 12 above)
Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: fixed

File: tests/python/unittest/test_numpy_op.py (hunk 13 above)
Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: fixed all @haojin2

Comment (hzfan) [reply tests/python/unittest/test_numpy_op.py]: As there are new commits in numpy branch, I rebase and force push.

**Comment by haojin2** [comment]
@hzfan LGTM, please resolve the conflicts and re-trigger CI

**Comment by haojin2** [comment]
@hzfan Please also add docs.

**Comment by hzfan** [comment]
docs added, conflicts resolved. @haojin2

**Pull request merged**